Feature: Adding interactive n.explore() function

bobbyxng opened the pull request:
## Changes proposed in this Pull Request
- Added basic n.explore() function based on folium and geopandas to interactively explore the network.
![image](https://github.com/user-attachments/assets/2af8dceb-f19e-4a89-aa0c-a4430160c3d7)
- Function will detect automatically what components are available (e.g. buses, links, transformers, generators, storage_units, loads)
- Layer control can be expanded to select rendered components
- Full functionality of folium map, e.g. map can be exported to self-contained html.

Example:
```
n = pypsa.Network("network.nc")
map = n.explore()  # folium map stored to map
map  # Call map to show interactive map e.g. in jupyter notebook
map.save("map_export.html")  # to create self-contained html.
```


## Checklist

- [X] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [X] Newly introduced dependencies are added to `environment.yaml`,  (folium is pretty heavy, should be installed manually).
- [X] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [X] I consent to the release of this PR's code under the MIT license.

lkstrp commented on pypsa/plot.py:
```diff
@@ -1233,3 +1243,160 @@ def iplot(
+        n.buses, geometry=gpd.points_from_xy(n.buses.x, n.buses.y), crs=crs
+    )
+
+    n.links = n.links[n.links.carrier == "DC"]  # Filter out non DC links
```
You are overwriting the network links here. This will break the network object. Use a temporary filtered list

bobbyxng replied:
Good point, corrected!

lkstrp commented on pypsa/plot.py:
```diff
@@ -1233,3 +1243,160 @@ def iplot(
+        components_present.append("links")
+
+    if not gdf_buses.empty:
+        buses_group = FeatureGroup(name="Buses")
```
This is unused

bobbyxng replied:
Thanks, relict of a feature I have not yet implemented :D

lkstrp commented on pypsa/plot.py:
```diff
@@ -33,6 +36,13 @@
+    import mapclassify
+except ImportError:
+    explore_deps_present = False
+
```
Since we only use those dependencies for n.explore, I think we can move the import in the function

bobbyxng replied:
good point, done!

lkstrp commented on pypsa/plot.py:
```diff
@@ -1233,3 +1243,160 @@ def iplot(
+        )
+        components_present.append("buses")
+
+    logger_explore = ""
+    if len(components_present) > 0:
+        logger_explore += (
+            f"Components rendered on the map: {', '.join(components_present)}."
+        )
+    if len(set(components) - set(components_present)) > 0:
+        logger_explore += f" Components omitted as they are missing: {', '.join(set(components) - set(components_present))}."
+    logger.info(logger_explore)
```
Just use logger.info directly here. You don't need logger_explore as an interim step

bobbyxng replied:
great, thanks!

lkstrp reviewed (changes requested):
The map is beauuutiful 💃

bobbyxng commented:
@lkstrp @fneum @FabianHofmann : I implemented all comments. Further I added more components that can be explored, if they are present: 

```
"buses",
"lines",
"links",
"transformers",
"generators",
"loads",
"storage_units",
```

From my side, basic functionalities are now provided and checked.

bobbyxng commented:
Added testing routing to ```test_plotting.py```, thanks @lkstrp !
```
try:
    import folium
    import mapclassify

    explore_deps_present = True
except ImportError:
    explore_deps_present = False
```
```
@pytest.mark.skipif(not explore_deps_present, reason="Dependencies for n.explore() not installed: folium, mapclassify")
def test_network_explore(ac_dc_network):
    n = ac_dc_network

    map = n.explore()
```

```

bobbyxng commented:
Inspiration for future functionalities:
- https://deepnote.com/app/allan-campopiano/Folium-and-Charts-01298f69-681b-4c83-9487-2e9e2a74e786
- https://python-visualization.github.io/folium/latest/advanced_guide/piechart_icons.html

lkstrp commented:
Folium looks pretty amazing.

@fneum @bobbyxng 
What do we think about enhancing this further and use it as a full replacement for `n.plot()` and `n.iplot()`? I guess, we can still create static plots from it and recreate full functionality. And so much more so easily. I nice fresh rebuild would also solve most of the mess in `plot.py`.

fneum commented:
`iplot()` can be replaced, but `plot()` cannot (we need it for paper plots :) )

bobbyxng commented:
> Folium looks pretty amazing.
> 
> @fneum @bobbyxng What do we think about enhancing this further and use it as a full replacement for `n.plot()` and `n.iplot()`? I guess, we can still create static plots from it and recreate full functionality. And so much more so easily. I nice fresh rebuild would also solve most of the mess in `plot.py`.

Definitely agree to push this further and enhancing it with plotting capabilities for solved networks (pie charts, on click information, etc.). However I do not see it as a replacement for `n.plot()` (for `n.iplot()` however I agree, as it is much faster and more resource-efficient), as n.plot() is still useful for pdf/image exports. `n.explore()` could serve, as the method name currently implies, for analytical purposes when assessing the network/results.

fneum commented:
@bobbyxng, looks amazing! Exactly what we need to debug networks :1st_place_medal: 

I have made some suggestions in #1011 to streamline the performance. It's easiest to go through them one by one as they involve some rearrangement of code.

1. Only use one tile, which can be passed as keyword argument.
2. Vectorize the LineString and Point creation.
3. Added keyword argument to select subset of existing components. Default are only transmission components.
4. And then, only compute GeoDataFrame for requested components.

For PyPSA-Eur `base.nc` this roughly cut ``%%timeit`` in half.

fneum commented on doc/references/release-notes.rst:
```diff
@@ -11,7 +11,7 @@ Upcoming Release
   ``pip install git+https://github.com/pypsa/pypsa#egg=pypsa``.
 
-* new feature or bugfix
+* Added basic n.explore() function based on folium and geopandas to interactively explore the network.
```
```suggestion
* Added basic ``n.explore()`` function based on folium and geopandas to explore the network interactively.
```
Split infinitive :book:

bobbyxng replied:
Thanks!

bobbyxng commented:
> @bobbyxng, looks amazing! Exactly what we need to debug networks 🥇
> 
> I have made some suggestions in #1011 to streamline the performance. It's easiest to go through them one by one as they involve some rearrangement of code.
> 
>     1. Only use one tile, which can be passed as keyword argument.
> 
>     2. Vectorize the LineString and Point creation.
> 
>     3. Added keyword argument to select subset of existing components. Default are only transmission components.
> 
>     4. And then, only compute GeoDataFrame for requested components.
> 
> 
> For PyPSA-Eur `base.nc` this roughly cut `%%timeit` in half.
> 
> The PR #1011 targets your feature branch, so merge at will.

Great improvements, much more efficient, merged! :)
